### PR TITLE
[Merged by Bors] - feat(Matrix): characteristic polynomial of upper triangular matrix

### PR DESCRIPTION
**Pull request by b-mehta**
Add simple lemmas relating block triangular matrices with characteristic polynomials.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/Block.lean [hunk 1]
```diff
@@ -6,6 +6,7 @@ Authors: Johannes Hölzl, Patrick Massot, Casper Putz, Anne Baanen, Wen Yang
 import Mathlib.LinearAlgebra.Matrix.Transvection
 import Mathlib.LinearAlgebra.Matrix.NonsingularInverse
 import Mathlib.Tactic.FinCases
+import Mathlib.Order.Extension.Linear
```
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/Block.lean:9]:
```suggestion
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 2]
```diff
@@ -112,6 +118,19 @@ lemma charpoly_fromBlocks_zero₂₁ :
...
     det_fromBlocks_zero₂₁]
 
+lemma toSquareBlock_charmatrix {α : Type*} [DecidableEq α] {b : n → α} {a : α}  :
+    (M.toSquareBlock b a).charmatrix = M.charmatrix.toSquareBlock _ _ := by
+  ext i j : 1
+  simp [charmatrix_apply, toSquareBlock_def, diagonal_apply, Subtype.ext_iff]
+
+lemma BlockTriangular.charpoly {α : Type*} {b : n → α} [LinearOrder α] (h : M.BlockTriangular b) :
+    M.charpoly = ∏ a ∈ image b univ, (M.toSquareBlock b a).charpoly := by
+  simp only [Matrix.charpoly, h.charmatrix.det, toSquareBlock_charmatrix]
```
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:128]:
```suggestion
  simp only [charpoly, h.charmatrix.det, toSquareBlock_charmatrix]
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 3]
```diff
@@ -112,6 +118,19 @@ lemma charpoly_fromBlocks_zero₂₁ :
   simp only [charpoly, charmatrix_fromBlocks, Matrix.map_zero _ (Polynomial.C_0), neg_zero,
     det_fromBlocks_zero₂₁]
 
+lemma toSquareBlock_charmatrix {α : Type*} [DecidableEq α] {b : n → α} {a : α}  :
+    (M.toSquareBlock b a).charmatrix = M.charmatrix.toSquareBlock _ _ := by
```
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:122]:
```suggestion
    (M.toSquareBlock b a).charmatrix = M.charmatrix.toSquareBlock b a := by
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 4]
```diff
@@ -112,6 +118,19 @@ lemma charpoly_fromBlocks_zero₂₁ :
...
     det_fromBlocks_zero₂₁]
 
+lemma toSquareBlock_charmatrix {α : Type*} [DecidableEq α] {b : n → α} {a : α}  :
+    (M.toSquareBlock b a).charmatrix = M.charmatrix.toSquareBlock b a := by
+  ext i j : 1
+  simp [charmatrix_apply, toSquareBlock_def, diagonal_apply, Subtype.ext_iff]
+
+lemma BlockTriangular.charpoly {α : Type*} {b : n → α} [LinearOrder α] (h : M.BlockTriangular b) :
+    M.charpoly = ∏ a ∈ image b univ, (M.toSquareBlock b a).charpoly := by
+  simp only [charpoly, h.charmatrix.det, toSquareBlock_charmatrix]
```
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:128]:
```suggestion
  simp only [Matrix.charpoly, h.charmatrix.det, toSquareBlock_charmatrix]
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 5]
```diff
@@ -112,6 +118,19 @@ lemma charpoly_fromBlocks_zero₂₁ :
   simp only [charpoly, charmatrix_fromBlocks, Matrix.map_zero _ (Polynomial.C_0), neg_zero,
     det_fromBlocks_zero₂₁]
 
+lemma toSquareBlock_charmatrix {α : Type*} [DecidableEq α] {b : n → α} {a : α}  :
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:121]:
```suggestion
lemma toSquareBlock_charmatrix {α : Type*} [DecidableEq α] {b : n → α} {a : α} :
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 6]
```diff
@@ -112,6 +118,19 @@ lemma charpoly_fromBlocks_zero₂₁ :
...
+    (M.toSquareBlock b a).charmatrix = M.charmatrix.toSquareBlock b a := by
+  ext i j : 1
+  simp [charmatrix_apply, toSquareBlock_def, diagonal_apply, Subtype.ext_iff]
+
+lemma BlockTriangular.charpoly {α : Type*} {b : n → α} [LinearOrder α] (h : M.BlockTriangular b) :
+    M.charpoly = ∏ a ∈ image b univ, (M.toSquareBlock b a).charpoly := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:127]: Is `image _ univ` not usually spelled `range`?

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 7]
```diff
@@ -4,6 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Kim Morrison
 -/
 import Mathlib.LinearAlgebra.Matrix.Adjugate
+import Mathlib.LinearAlgebra.Matrix.Block
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:7]: This seems to be somewhat heavy; I won't insist on it, but please consider whether it's worth putting these results in a new file.

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean (hunk 6 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:127]: `Set.image _ Set.univ` is, but not `Finset.image _ Finset.univ`.

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 8]
```diff
@@ -85,6 +86,11 @@ lemma charmatrix_fromBlocks :
   simp only [charmatrix]
   ext (i|i) (j|j) : 2 <;> simp [diagonal]
 
+lemma BlockTriangular.charmatrix {α : Type*} [Preorder α] {M : Matrix n n R} {b : n → α}
+    (h : M.BlockTriangular b) :
+    M.charmatrix.BlockTriangular b :=
+  (blockTriangular_diagonal _).sub fun i j h' => by simpa using h h'
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:92]: Is the reverse direction also true?

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean (hunk 5 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:121]:
```suggestion
lemma charmatrix_toSquareBlock {α : Type*} [DecidableEq α] {b : n → α} {a : α} :
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean (hunk 7 above)
Comment (b-mehta) [reply Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:7]: Hmm, I'm not convinced that there's enough to say about block triangular and characteristic matrices to deserve a new file, and it'd be much heavier to put these things in the block matrix file. @eric-wieser what do you think?

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean (hunk 5 above)
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:121]:
```suggestion
lemma charmatrix_toSquareBlock {α : Type*} [DecidableEq α] {b : n → α} {a : α} :
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean [hunk 9]
```diff
@@ -112,6 +118,19 @@ lemma charpoly_fromBlocks_zero₂₁ :
...
     det_fromBlocks_zero₂₁]
 
+lemma charmatrix_toSquareBlock {α : Type*} [DecidableEq α] {b : n → α} {a : α} :
+    (M.toSquareBlock b a).charmatrix = M.charmatrix.toSquareBlock b a := by
+  ext i j : 1
+  simp [charmatrix_apply, toSquareBlock_def, diagonal_apply, Subtype.ext_iff]
+
+lemma BlockTriangular.charpoly {α : Type*} {b : n → α} [LinearOrder α] (h : M.BlockTriangular b) :
+    M.charpoly = ∏ a ∈ image b univ, (M.toSquareBlock b a).charpoly := by
+  simp only [Matrix.charpoly, h.charmatrix.det, toSquareBlock_charmatrix]
```
Comment (b-mehta) [inline Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:128]:
```suggestion
  simp only [Matrix.charpoly, h.charmatrix.det, charmatrix_toSquareBlock]
```

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean (hunk 7 above)
Comment (b-mehta) [reply Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean:7]: Oof the import is indeed heavier than I thought. I still don't expect there's enough material about these to deserve a new file, but I'll add a comment suggesting that it should be split out if something else is thought of in the future. How does that sound to you?

File: Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean (hunk 8 above)
Comment (b-mehta) [reply Mathlib/LinearAlgebra/Matrix/Charpoly/Basic.lean]: Added the iff and reverse directions. It suggested four more lemmas about block triangular matrices, so I've added them to that file.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**